File: 프로그래머스/lv4/43236. 징검다리/징검다리.py
def solution(distance, rocks, n):
    rock_positions = sorted(rocks)
    
    def cond(s,n):
        # 바위를 거리의 최소값이 s 일때 바위를 n 개이상 지울 수 있는지 여부 반환
        removed = 0
        tmp_position = 0
        for rock_position in rock_positions:
            if rock_position-tmp_position < s:
                removed+=1
                if removed > n:
                    return True
            else:
                tmp_position = rock_position

        return False

    def bisect(left ,right, cond):
        while left <= right:
            mid = (left+right)>>1
            if cond(mid,n): 
                right = mid - 1
            else: 
                left = mid + 1
        return right
    
    return bisect(0, distance, cond)

File: 프로그래머스/lv4/43236. 징검다리/징검다리.js
function solution(distance, rocks, n) {
  rocks = rocks.sort((a,b) => a-b)
  rocks.push(distance)
  return bisect_l(0,distance,cb,rocks,n)
}

function cb(dists,n,maxDist) {
  let pos = 0
  for (let dist of dists) {
    if (dist-pos<maxDist) {
      n-=1
      if (n<0) return false
    }
    else pos = dist
  }
  return true
}

function bisect_l(l,r,cb,dists,n) {
  while (l<=r) {
    mid = (l+r)>>1
    if (cb(dists,n,mid)) {l = mid+1}
    else {r = mid-1}
  }
  return r
}